src/sql/src/plan/lowering.rs

```diff
@@ -455,6 +458,7 @@ impl HirRelationExpr {
                                 requires_nonexistent_column
                             })
                             .unwrap_or(scalars.len());
+                        assert!(end_idx > 0, "a Map expression references itself or a later column; lowered_arity: {}, expressions: {:?}", lowered_arity, scalars);
```

Comment:
Useful independently. Without the assert, if there is a bug in the preparation of the HIR expressions such that this condition ends up false, then we get into an infinite loop here, which would be really bad in production. It's much less bad to crash instead. (Without this assert, an envd thread would be spinning forever, and then it would be really hard to debug what went wrong if this happens in production. But even when it happens locally it makes it easier to figure out what's going on.)

During the development of this PR, I twice had bugs where I got into an infinite loop here. I can imagine similar bugs happening in other situations later.